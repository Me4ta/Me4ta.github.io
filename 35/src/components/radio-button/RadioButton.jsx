import React from 'react'
import classNames from 'classnames'

import './RadioButton.scss'

export default class RadioButton extends React.Component {
	constructor(props) {
		super(props)

		this.state = { checked: this.props.checked }
	}

	componentWillReceiveProps(nextProps) {
		this.setState({ checked: nextProps.checked })
	}

	shouldComponentUpdate(nextProps, nextState) {
		return this.state.checked !== nextState.checked || nextProps.checked !== this.state.checked
	}

	handleClick = e => {
		e && e.preventDefault()

		this.setState({ checked: !this.state.checked }, () => {
			this.state.checked ? this.props.on() : this.props.off()
		})
	}

	render() {
		const { checked } = this.state
		const className = classNames('radio-button', this.props.className, { '_active': checked })

		return (
			<label className={className} onClick={this.handleClick}>
				<input type='checkbox' name={this.props.name} checked={Boolean(checked)} />
			</label>
		)
	}
}

RadioButton.propTypes = {
	checked: React.PropTypes.bool,
	name: React.PropTypes.string,
	className: React.PropTypes.string,
	on: React.PropTypes.func,
	off: React.PropTypes.func
}

RadioButton.defaultProps = {
	checked: false,
	name: '',
	className: '',
	on: () => {},
	off: () => {}
}
