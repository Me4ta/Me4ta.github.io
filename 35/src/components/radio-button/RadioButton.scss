@import "~style/variables.scss";
@import "~style/lib/mixins.scss";

.radio-button {
	display: inline-block;
	width: 5rem;
	height: 2.5rem;
	border-radius: 1.25rem;
	background-color: #999;
	cursor: pointer;
	transition: background-color $transitionSpeed;
	position: relative;
	//vertical-align: middle;

	&:before {
		content: '';
		position: absolute;
		@include lt(0.1rem, 0.1rem);
		@include wh(2.3rem, 2.3rem);
		background: white;
		border-radius: 100%;
		transition: left $transitionSpeed;
	}

	&._active {
		background-color: $brightColor;

		&:before {
			left: calc(100% - 2.4rem);
		}
	}

	INPUT[type="checkbox"] {
		display: none;
	}
}