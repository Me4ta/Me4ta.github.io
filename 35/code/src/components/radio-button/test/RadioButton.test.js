import React from 'react'
import RadioButton from '../RadioButton'
import renderer from 'react-test-renderer'

it('renders correctly', () => {
	const component = renderer.create(
		<RadioButton checked={Boolean(true)} name='test' className='additional-class' />
	)

	let tree = component.toJSON()
	expect(tree).toMatchSnapshot()

	tree.props.onClick()
	tree = component.toJSON()
	expect(tree).toMatchSnapshot()
})