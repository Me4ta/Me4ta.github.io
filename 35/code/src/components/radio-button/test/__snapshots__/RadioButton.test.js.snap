exports[`test renders correctly 1`] = `
<label
  className="radio-button additional-class _active"
  onClick={[Function]}>
  <input
    checked={true}
    name="test"
    type="checkbox" />
</label>
`;

exports[`test renders correctly 2`] = `
<label
  className="radio-button additional-class"
  onClick={[Function]}>
  <input
    checked={false}
    name="test"
    type="checkbox" />
</label>
`;
